### PR TITLE
[NOT-FOR-MASTER: Obsolete] Operator versioning proposal. See terminal comment for status.

### PR DESCRIPTION
**Pull request by ezyang**
Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: docs/Versioning.md [hunk 1]
```diff
@@ -55,11 +55,142 @@ By way of example, the `ModelProto.ir_version` MUST be present in every model.
...
+with some machine-readable metadata which specifies the implementation
+for every operator that is in scope.  When ONNX makes BC-breaking
+changes to the semantics of operators (e.g., by changing the meaning
+of an operator or removing an operator), it increments the ONNX
+operator version.  Additions of new operators do NOT increment the
+ONNX operator version (as they are not BC-breaking changes), nor
+do additions of new optional attributes.
+
+When understanding the semantics of an operator version, it is helpful
+to distinguish an *operator implementation* from an operator name
```
Comment (linkerzhang) [inline docs/Versioning.md]: Change "operator implementation" to "operator definition" or "operator specification"? For "operator implementation", it means "kernel codes" normally?

File: docs/Versioning.md [hunk 2]
```diff
@@ -55,11 +55,142 @@ By way of example, the `ModelProto.ir_version` MUST be present in every model.
...
+of an operator (its type, the mathematical computation it performs on
+its arguments, etc.); an operator name is simply what is recorded in an
+exported model.  Operator implementations are *immutable*; they never
+change across ONNX operator versions.  For example, if ONNX wants to
+introduce a breaking change to the `Conv` operator, it must create a
+new operator implementation describing the new desired behavior, and
+then remap the `Conv` name to this new implementation in the next ONNX
+operator version.  In this way, the ONNX operator version is a
+*namespace* mechanism on top of an immutable store of operator
+implementations; each operator version defines a mapping of operator
```
Comment (linkerzhang) [inline docs/Versioning.md]: My understanding of this design is, the operator implementation store will never be cleaned up. 

One personal thinking please. I'm in favor of having the capability to clean up operator implementation store. The operator implementation store is part of onnx standard.  We may want to clean up the standard when we feel too old versions do not make sense and we want to enforce an upgrade of vendors' components (importers/exporters). Without the clean up capability, it's always up to vendors to decide the range of operator implementation store they'll support. Does this make sense please?

File: docs/Versioning.md [hunk 3]
```diff
@@ -7,13 +7,13 @@ specification, MUST, SHOULD et al are used consistent with [RFC2119](https://too
 
 ONNX defines versioning policy and mechanism for three classes of entities:
 
-* The abstract graph model and the concrete format that serializes it. The graph model and format are versioned atomically and are referred to as the *IR version.* The IR version is  represented by the `ModelProto.ir_version` field.
-* Operators that are referenced by a given ONNX graph. The version of a given operator  is referred to as the *operator version*. The operator version is  represented by the `TBD` field.
+* The abstract graph model and the concrete format that serializes it. The graph model and format are versioned atomically and are referred to as the *IR version.* The IR version is represented by the `ModelProto.ir_version` field.
+* Operators that are referenced by a given ONNX graph.  The set of ONNX operators in scope and their semantics is versioned via the *ONNX operator version*. The ONNX operator version is represented by the `ModelProto.op_version` field.
```
Comment (prasanthpul) [inline docs/Versioning.md]: we should clarify whether ONNX and ONNX-ML (and any other profile) have independent versioning for IR and operators

File: docs/Versioning.md (hunk 2 above)
Comment (ezyang) [reply docs/Versioning.md]: The proposal, as I understand it, is compatible with what you want. The point is that the operator version standard doesn't *mandate* that an ONNX importer actually have implementations for all operator implementations in the store; they can do whatever they want. We can give guidelines about what we recommend they implement fully to make this clearer.

File: docs/Versioning.md (hunk 3 above)
Comment (ezyang) [reply docs/Versioning.md]: I think that it must version separately, it doesn't work reasonably otherwise.

Comment (prasanthpul) [reply docs/Versioning.md]: then we need a field in ModelProto to specify the profile

**Comment by ezyang** [comment]
I was thinking about how to simplify this proposal more, and I had (what I think is) a good idea: what if we get rid of the "operator definition" name (the thing that has a leading underscore in my proposal), and instead uniquely identify operators using the operator name (`op_type`) and the *first ONNX operator version they were introduced.* I propose the syntax is `OpName:GlobalOpVer`, e.g., `Add:5` or `Conv:27`.

This is NOT introducing per operator versions, because the "version" tagged on each operator reflects the global operator version, e.g., it's very unlikely you'll have `Add:1`, `Add:2`, `Add:3`, rather, it will look more like `Add:1`, `Add:55`, `Add:123`. You get the point. And like before, operator definitions are still immutable, so you can always talk about `Add:55` and it will always be the same implementation of `Add` it always was.

I think there a few merits to this proposal:

1. It reduces the number of concepts (no more leading underscore names)
2. It reduces the amount of naming we ONNX specifiers have to do (no more having to pick a new operator definition name for conv with bias, the numbering is mandated!)
3. It makes it clearer how to write operator definitions; an operator definition looks as it does today, but we add a new field "Since" which records the earliest ONNX operator version which shipped with this operator (conveniently, this coincides with the tag!)
4. We need ONLY record operator names in the  "Added/Modified/Removed" fields in the changelog, since you can infer the operator version number that was added based on the version the changelog entry corresponds to.

File: docs/Versioning.md (hunk 1 above)
Comment (donbox) [reply docs/Versioning.md]: I would prefer that we simply snap to CS conventions and leverage that intuition.

Operators are declared - their signature and semantics are specified, but their implementation is opaque.
Graphs are defined - their signature and semantics are specified, and their implementation is transparent.

This is consistent with other language specs. In C:
     extern float add(float a, float b);              // declares add as a name that can be referenced
     float add(float a, float b) { return a + b; } // defines the implementation of add 

I prefer declaration, but am fine with specification or contract. Definition implies a specific implementation, which I'm not a fan of.

**Comment by ezyang** [comment]
> As to whether we call these things namespaces, libraries, modules, or packages, I'm not married to any one term. To avoid picking a name, I will use LNMP throughout my comments as a placeholder for whichever one we land on.

I think in the spec text, I used the term "vendor (operator) extension", to make it clear that the third parties that are most likely to define extra operators are the backend vendors, who would actually be in the business of implementing them. The reason I don't like library/module/package is because it brings the intuition that *users* could write libraries of operators, composed out of more primitive ones, that could portably be reused over many backends (and then you'd have some sort of package manager, etc. etc. etc.) This might be a legitimate problem to solve, but it's not the one we're trying to solve with operator versioning.

> I prefer declaration, but am fine with specification or contract. Definition implies a specific implementation, which I'm not a fan of.

I'm absolutely fine with declaration/specification. They are *solely* called definitions right now, because that is the internal naming convention in ONNX library. I can rename `def` to `spec` in the library to make this clear.

**Comment by donbox** [comment]
Edward and I worked through my comments today and we have a new, simpler design that I am writing up.

We agreed to do it in a separate branch and PR. Look for the prose and proto tonight.

Until then, please hold you commentary here until the new PR is in place.

**Pull request closed without merging**